A method to print edge energy interactively

ptim0626 opened the pull request:
This is related to issue #2409. Any suggestion is welcomed.

### Description of the change
An option is provided in `print_edges_near_energy()` to print the table interactively, or the method `interactive_edges_energy()` can be called directly to achieve this.

An option is also added in `print_edges_near_energy()` to show only the major edges if desired.

### Progress of the PR
- [x] Change implemented,
- [x] update docstring,
- [x] update user guide,
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
![iss2409_sc](https://user-images.githubusercontent.com/62692924/84412076-a10d3500-ac06-11ea-996b-1b7dffd8ad13.png)

ptim0626 committed:
Add a method to print edge energy interactively

An option is provided in print_edges_near_energy to print
the table interactively, or the method interactive_edges_energy
can be called directly to achieve this.

An option is also added in print_edges_near_energy to
show only the major edges if desired.

francisco-dlp commented:
That's cool! Could you move the widgets code to https://github.com/hyperspy/hyperspy_gui_ipywidgets? HyperSpy should not depend on ipywidgets.

ericpre commented:
Nice indeed, for the interactivity part, an alternative may be something along the line what does the `calibrate` method:
![Peek 2020-06-11 18-16](https://user-images.githubusercontent.com/11851990/84420668-48dc3000-ac12-11ea-8718-cb9a21b8fab5.gif)

francisco-dlp commented:
@ericpre, that's ultracool :) In this way one can select an energy window an move it around: I think that there is no better in terms of usability but maybe somebody will surprise us.

francisco-dlp commented:
A small suggestion: instead of the interactive keyword, it could default to it if no arguments are passed to the function.

ericpre commented:
Having it for EDS signal would be very good too! :)

@ptim0626, please let us know if you need help to find your with the widget registration mechanism, etc.
The `calibrate` widget and method should be a good place to start: it is a bit convoluted and it involves [widgets in hyperspy_gui_ipywidgets](https://github.com/hyperspy/hyperspy_gui_ipywidgets/blob/e18aca582899bd9ba94cb74bf75da9c04bda5d52/hyperspy_gui_ipywidgets/tools.py#L67-L148) + [this](https://github.com/hyperspy/hyperspy_gui_ipywidgets/blob/e18aca582899bd9ba94cb74bf75da9c04bda5d52/hyperspy_gui_ipywidgets/hyperspy_extension.yaml#L33) + [this](https://github.com/hyperspy/hyperspy/blob/deff3fb05a9ab9e7325199aab43861a229f81fd2/hyperspy/hyperspy_extension.yaml#L246) and [this](https://github.com/hyperspy/hyperspy/blob/df2af64ae8e8e077e967ae2a9d76a3539beb47ed/hyperspy/_signals/signal1d.py#L854)! ;)



> A small suggestion: instead of the interactive keyword, it could default to it if no arguments are passed to the function.

Here is an example:
https://github.com/hyperspy/hyperspy/blob/df2af64ae8e8e077e967ae2a9d76a3539beb47ed/hyperspy/_signals/signal1d.py#L784-L800

francisco-dlp commented:
Another usability suggestion that is more involved: select an edge from the list and display vertical line markers at the position of all edges of that element in spectrum energy range—that's the usual way to identify edges when in doubt.

francisco-dlp commented:
A simpler variation (that may even be better usability-wise) would be to have an option (with the check button) to display markers at the position of all the edges in the *spectrum range* of all the elements that have edges in the selected *energy range*.

ptim0626 commented:
@ericpre Thanks for pointing me to the ```calibrate()``` method, I shall have a good look at it and incorporate my method similarly.
@francisco-dlp I think it would be nice to have all those options for users, will see how far I can go (perhaps a bit too ambitious...)

ericpre commented:
> @ericpre Thanks for pointing me to the `calibrate()` method, I shall have a good look at it and incorporate my method similarly.
> @francisco-dlp I think it would be nice to have all those options for users, will see how far I can go (perhaps a bit too ambitious...)

Indeed, some of it can be done later in separate PR.
On the topic of displaying edge, there is this PR: https://github.com/hyperspy/hyperspy/pull/1527

ptim0626 committed:
Refactor printing edges table interactively

A method edges_at_energy is added so that it uses the registered
widget for printing edges table interactively.

ptim0626 committed:
Refactor the utility function get_edges_near_energy

Add options for sorting the edges.

ptim0626 committed:
Add options for sorting onset energy in printed table

This includes 'closest', 'ascending' and 'descending'.

ptim0626 committed:
Put vertical lines and labels when edges are selected

ptim0626 committed:
Improve the placement of edge labels

ptim0626 committed:
Add methods to show and plot complementary edges

ptim0626 committed:
Handle placement of labels when figure is resized

ptim0626 committed:
Refactor some parts of the EdgeRange class

ptim0626 committed:
Handle None span selector

ptim0626 commented:
I have moved the ipywidget part to hyperspy/hyperspy_gui_ipywidgets#28. Now an energy range can be selected in the spectrum and it will display edges within the range, the way of the displayed edges can be changed, such as including only major edges. Upon selecting the edges, it will add markers to the signal to indicate their positions and an option is added so that edges of the same element in the energy axis range can be shown (useful for identifying edges). Here is a typical usage:
![hyperspy_iss2409](https://user-images.githubusercontent.com/62692924/85123770-c01c5000-b220-11ea-9efb-e80d40ff5b8b.gif)

On the matter of displaying edges, I do not include those in the EELS signal class and also not sure the status of #1527. My current edge label plotting tries to spread them out and does not explicitly check if any overlap occurred. If desired, I can refactor the relevant part to the EELS signal class and further improve the edge label plotting mechanism in a separate PR.

ericpre reviewed (commented):
This is already fairly nice. The code adding label on the figure should go in the EELS signal class, ideally most of it should go in a separate class so it can be also used for EDS signal but also other type of signal. Currently this PR doesn't support blitting, which doesn't matter with the notebook backend (since it doesn't support blitting) but for the qt backend, it can be significantly faster.

I think this could be simplify if following is possible:
- add tickboxes for each row in the table to display the label on the figure - I don't know if this is easily feasible but it will be easier to relate the selected/displayed edge and the corresponding information in the table
- when selecting a edge to plot, display all edges by default
- remove the "show_edge(s)" and "other edge(s)" selection widgets

The only advantage of the two selection widgets (in comparison to the tickbox approach) is that it is possible to select several at the same time but usually not many edges are relevant and we inspect these one by one, mainly because of the overlap issue.

There should be some speed issue somewhere because I would expect this to be faster.

If you want to keep this structure for this PR, then we can review it and the rest can be done in other PRs. Personally, I would prefer to have the label plotting done sorted first and then come back to this but this is up to you!

ptim0626 commented:
@ericpre I will see what I can do for selecting edges, the reason I use those selection widget is that there seems to be no easy way to select entries in the table directly and I don't know how easy or difficult to align tickboxes to the rows of table accurately. I will see what I can do for this. And yes, I think should refactor the label plotting mechanism to a separate class, so both EELS and x-ray lines spectrum can utilise it.

ptim0626 committed:
Refactor the edge labelling into a new class

The new class SpectrumLabelPosition provides markers for the EELS
signal, also potentially for EDS. A plot method is added in EELS
signal class, which works off from #1527.

ptim0626 committed:
Avoid running get_edge_near_energy everytime when selector changes

get_edge_near_energy needs to loop over all elements every time.

ptim0626 committed:
Rewrite some implementation of EdgesRange for GridBox table format

ptim0626 commented:
I implement the `plot` method in EELS as well, basically from the idea of #1527. The edge labelling mechanism is refactored into a new class. In term of scope I think this PR covers enough already. I emulate a 'clickable table' by assembling widgets into a grid and I add a button to update the table when it is told to do so it is smoother. I can add some tests and docs if this skeleton is ok.
![hyperspy_iss2409_2](https://user-images.githubusercontent.com/62692924/86057115-74ef1200-ba56-11ea-861e-1413d27765de.gif)

thomasaarholt commented:
Great! This looks really cool! I'll check it out this week if I get the chance!

francisco-dlp commented:
@ptim0626, before I forget it, could you update your name and institution in your GitHub profile. In this way they'll appear in the DOI that is automatically generated by zenodo using GitHub metadata.

ericpre commented on hyperspy/misc/label_position.py:
```diff
@@ -0,0 +1,220 @@
+import itertools
+
+import numpy as np
+from hyperspy.misc.eels.tools import get_edges_near_energy, get_info_from_edges
```
Remove the `get_edges_near_energy` import, which is most likely a left over from previous version?

ptim0626 replied:
Yes it is

ericpre commented on hyperspy/misc/label_position.py:
```diff
@@ -0,0 +1,220 @@
+from hyperspy.drawing.marker import markers
+
+class SpectrumLabelPosition():
+    edge_label_style = {'ha' : 'center', 'va' : 'center', 
```
Add a short description of what it does and how it is expected to be used.

ptim0626 replied:
Added

ericpre commented on hyperspy/_signals/eels.py:
```diff
@@ -1433,6 +1487,197 @@ def create_model(self, ll=None, auto_background=True, auto_add_edges=True,
+            edges = self._get_edges_to_plot(include_edges, only_edges)
+            self.put_label_on_signal(edges)
+
+    def put_label_on_signal(self, edges, vertical_line_marker=None, 
```
Rename to `plot_edges_label` to be more consistent with other similar method (`plot_markers`, `_plot_xray_lines`)?

ptim0626 replied:
Renamed to `plot_edges_label`

ericpre commented on hyperspy/_signals/eels.py:
```diff
@@ -1433,6 +1487,197 @@ def create_model(self, ll=None, auto_background=True, auto_add_edges=True,
+        # add the markers to the signal and store them
+        self.add_marker(vertical_line_marker + text_marker)
+        info = dict(zip(edges, zip(vertical_line_marker, text_marker)))
+        self._edge_markers = {k:list(v) for k, v in info.items()}
```
Is the `info` dictionary necessary? Alternatively, do the dictionary values need to be a list and not a tuple?
```suggestion
        self._edge_markers = {k:list(v) for k, v in zip(edges, zip(vertical_line_marker, text_marker))}
```

ptim0626 replied:
Used `map` to change the zip's tuple to list in one go now. The values of the dictionary (which stores the markers) needs to be a list to allow popping and removing of the markers.

ericpre commented on hyperspy/_signals/eels.py:
```diff
@@ -1433,6 +1487,197 @@ def create_model(self, ll=None, auto_background=True, auto_add_edges=True,
+            try:
+                elements = self.metadata.Sample.elements
+            except AttributeError:
+                raise Warning("No elements defined. Add them with "
```
I would use a `ValueError` instead of a warning, since it doesn't make sense to use `include_edges=True` if the elements are not defined.

ptim0626 replied:
Agreed.

ericpre commented on hyperspy/_signals/eels.py:
```diff
@@ -1433,6 +1487,197 @@ def create_model(self, ll=None, auto_background=True, auto_add_edges=True,
                           dictionary=dictionary)
         return model
 
+    def plot(self, include_edges=False, only_edges=('Major', 'Minor'), 
```
Rename `include_edges` to something more consistent with the rest of hyperspy? The corresponding argument for EDS signal is not a great example (`xray_lines`) but for model and more generally signal, we have `plot_components` and `plot_markers`, so I would suggest to use `plot_edges` or `plot_edges_label`. @hyperspy/core, any opinion?

For HyperSpy 2.0, it may be worth renaming `xray_lines` to something better.

ptim0626 replied:
Renamed to `plot_edges`

ericpre commented on hyperspy/signal_tools.py:
```diff
@@ -238,7 +243,162 @@ def apply(self):
+                                        vertical_line_marker=vm, 
+                                        text_marker=tm)
+        self.active_edges = active 
+        self.active_complementary_edges = complementary
```
Are these two lines necessary?

ptim0626 replied:
Removed

ericpre commented on hyperspy/signal_tools.py:
```diff
@@ -238,7 +243,162 @@ def apply(self):
+        if complementary is None:
+            complementary = self.active_complementary_edges
+            
+        self._clear_all_markers()
```
I am not sure removing all labels should be done here: this doesn't play well with the qt backend (labels are unnecessarily remove and re-added).

ptim0626 replied:
Oh I haven't checked the qt backend! The marker's positions are now updated instead of removing them first and then adding them, this gives a smoother experience.

ericpre commented on hyperspy/signal_tools.py:
```diff
@@ -238,7 +243,162 @@ def apply(self):
+        self.description_all = np.asarray(description_all)
+
+    def update_table(self):
+        figure_changed = self.slp._check_signal_figure_changed()
+        if figure_changed:
+            self._clear_all_markers()
```
When changing navigation index, drawing of the line is connected to the `axes_manager`: there should be a method which update the labels (their vertical position) when the nagivation indices change to take into account changes in intensity.
It may be worth looking at how some methods of the `BackgroundRemoval` tool are connected or how plotting x-rays lines works too.

ptim0626 replied:
The event `axes_manager.events.indices_changed` is connected to a method `_on_nav_change` now so that whenever the indices of `axes_manager` changed, it gets the new figure properties and update the existing labels.

ericpre commented on hyperspy/misc/label_position.py:
```diff
@@ -0,0 +1,220 @@
+import itertools
+
+import numpy as np
+from hyperspy.misc.eels.tools import get_edges_near_energy, get_info_from_edges
```
This file should be not be signal specific and generic enough to support different type of signal: would it be possible to `get_info_from_edges`?

ptim0626 replied:
Removed all dependency on `hyperspy.misc.eels.tools`

ericpre commented on hyperspy/_signals/eels.py:
```diff
@@ -1433,6 +1487,197 @@ def create_model(self, ll=None, auto_background=True, auto_add_edges=True,
+                line_markers = self._edge_markers[EELS_edge]
+                while line_markers:
+                    m = line_markers.pop()
+                    m.close()
```
It may be worth considering using the argument `render_figure` to speed up removing marker when using matplotlib backend supporting blitting but let sort out first the `_clear_all_markers` bit.

ptim0626 replied:
`render_figure` is set `False` now whenever markers are added or closed.

ericpre reviewed (changes requested):
Nice, this is very good progress, considering this isn't a straightforward PR - the review takes time too! ;)
The structure is already in good shape but there are a couple of things to improve regarding adding/removing markers and update on changing navigation indices.

The custom table with the button is nice. The only drawback with this approach is that it seems that generating the table is slower than generating the html and I guess it may too slow to have the table updating interactively with moving the span selector. However, this is already good enough and this could be improve later (maybe doing it asynchronously or something similar).

francisco-dlp commented:
@ericpre and @ptim0626, what's the status of this? It would be great to include it in this week's release.

ptim0626 commented:
@francisco-dlp sorry I was occupied by other things, will try my best to see if I can squeeze some time tomorrow to finish this off.

francisco-dlp commented:
Great, thanks!

ptim0626 committed:
Make the SpectrumLabelPosition class more general

Remove EELS specific methods/functions

ptim0626 committed:
Rename methods and simplify codes relating to EELS plotting

ptim0626 committed:
Replot edge labels when navigator changes

ptim0626 committed:
Update label positions instead of removing then adding them

Smoother in the qt backend.

ptim0626 committed:
Do not render figure when adding or removing markers for labels

ptim0626 committed:
Add tests for new methods in EELS signal class

ptim0626 committed:
Merge branch 'RELEASE_next_minor' into iss2409

ptim0626 committed:
Remove a test for interactive plotting for EELS

ptim0626 committed:
Add tests for utility EELS tools

ptim0626 committed:
Add documentation for the GUI for EELS edge identification

francisco-dlp commented:
@ericpre, do you think that this one can make it into v1.6?

ericpre commented on hyperspy/_signals/eels.py:
```diff
@@ -1433,6 +1487,211 @@ def create_model(self, ll=None, auto_background=True, auto_add_edges=True,
+        # add the markers to the signal and store them
+        self.add_marker(vertical_line_marker + text_marker, render_figure=False)
+        added = dict(zip(edges, map(list, zip(vertical_line_marker, text_marker))))
+        self._edge_markers.update(added)
```
When changing indices, the labels doesn't update, while they do with the interactive tool. I guess the function drawning/updating the label should be connected to the `axes_manager` here.

An example to reproduce:
```python
s = hs.datasets.artificial_data.get_core_loss_eels_line_scan_signal(True)
s.add_elements(['Cr', 'Fe'])
s.plot(plot_edges=True)
```

ptim0626 replied:
Thanks for checking this! completely overlooked... now the labels will update both in static and interactive mode.

ericpre commented on hyperspy/signal_tools.py:
```diff
@@ -244,6 +249,191 @@ def apply(self):
+        self._get_edges_info_within_energy_axis()
+
+        self.signal.axes_manager.events.indices_changed.connect(self._on_nav_change,
+                                                                [])
```
When connecting, we also need to connect the "disconnection" when closing. For example:

https://github.com/hyperspy/hyperspy/blob/3ec3018f59cee4c9db023408b09d51024fe53d42/hyperspy/signal_tools.py#L616-L621

ptim0626 replied:
now it connects the 'disconnection' when `signal_plot` is closed

ericpre reviewed (commented):
A few things need to be fixed but this is close to be finished. If this is possible to fix these tomorrow, it should be possible to include these in 1.6.

It would be good that the `SpectrumLabelPosition` class is also used for plotting X-rays line, but this could be done in another PR.

ptim0626 committed:
Disconnect the connected function when axes_manager's indices changed

ptim0626 committed:
Ensure labels are updating in both interactive and non-interactive mode

ericpre committed:
Increase test coverage.

ericpre committed:
Add plotting test.

ericpre commented:
Thanks @ptim0626 for the quick update. I have added tests in https://github.com/ptim0626/hyperspy/pull/1. It would be good to increase the coverage of the `EdgesRange` (https://codecov.io/gh/hyperspy/hyperspy/compare/db4dad25bb47850cbeef00c886def2d058241e55...ef64a958f5aac7cbc518097fa2b5a0c0514b4ebf/src/hyperspy/signal_tools.py)

ptim0626 committed:
Add tests for the EdgesRange class

ptim0626 committed:
Merge pull request #1 from ericpre/iss2409

Increase test coverage.
